Convert script to Swift Package Manager library

Samasaur1 opened the pull request:
This PR converts the standalone script to a Swift Package Manager library. This is primarily so that I can add this module to Nixpkgs, because the tooling is better for SPM packages than standalone scripts. This change also allowed me to remove the manual parsing of the SDK, which both makes the build similar and also makes it easier to package

Samasaur1 committed:
Convert to Swift Package Manager module

Samasaur1 committed:
Update Makefile to use SPM build

This also avoids the incredibly cursed SDK parsing

mostpinkest commented on Makefile:
```diff
@@ -2,25 +2,15 @@ VERSION = 2
-	swiftc watchid-pam-extension.swift $(DEFINES) -o $(LIBRARY_PREFIX)_x86_64.so -target x86_64-$(TARGET) -emit-library
-	swiftc watchid-pam-extension.swift $(DEFINES) -o $(LIBRARY_PREFIX)_arm64.so -target arm64-$(TARGET) -emit-library
-	lipo -create $(LIBRARY_PREFIX)_arm64.so $(LIBRARY_PREFIX)_x86_64.so -output $(LIBRARY_NAME)
+	swift build -c release --triple x86_64-$(TARGET)
```
Seen as the target is already specified in Package.swift, is there any reason why we're specifying it again here and not going straight to something like:

```
swift build -c release --arch x86_64 --arch arm64
```

Samasaur1 replied:
Oh yeah, that's way better. I didn't know that was an option

mostpinkest commented on Sources/pam-watchid/pam_watchid.swift:
```diff
@@ -80,7 +80,7 @@ private func parseArguments(argc: Int, argv: vchar) -> [String: String] {
 private extension LAPolicy {
     static var deviceOwnerAuthenticationIgnoringUserID: LAPolicy {
-#if SEQUOIASDK
+#if compiler(>=6.0)
```
Usually the swift used is the one installed as part of Xcode, but it's also possible to install a newer version of the toolchain manually from [swift.org](https://swift.org/). This would allow the 6.0 compiler to be used on older versions of macOS which don't support the required SDK. The current version of swift is being built to support versions of macOS as old as Big Sur. This causes errors like this:

<img width="977" alt="Screenshot 2024-12-20 at 11 57 54 pm" src="https://github.com/user-attachments/assets/b752973e-607c-47bf-a88f-5fbfab77593b" />

I know it's an absurdly niche environment, but it's technically a regression and not something I want to potentially have to deal with later.

As an alternative, `canImport()` could be used to check if a newer framework is available. For example [ImagePlayground](https://developer.apple.com/documentation/imageplayground) was added in macOS 15.1 according to the docs. I haven't had a good look so there might be something better suited but if not, I'm happy to set the cutoff at 15.1 as it should still be compatible.

Samasaur1 replied:
I was aware of that problem, yeah, but I couldn't find any frameworks that were introduced specifically in 15.0, and if I checked with ImagePlayground, then this package would fail to build on macOS 15.0..<15.1, which I thought was a bigger sacrifice to make than this niche scenario

Samasaur1 replied:
Maybe I could combine the checks?

Samasaur1 replied:
aha! [this site](https://marcoeidinger.github.io/appleframeworks/) was my savior, and I can check `#if canImport(CoreHID)`

mostpinkest replied:
Great find!

mostpinkest reviewed (commented):
Thanks for the pr. Sorry it's taken a bit for me to get around to responding.

I love the premise of Nix and am thankful you've taken the time to help get this working.

The Swift Package Manager aspects look great. Just have some concerns around the SDK version compatibility. If anything is needed in order to get it working with Nix, please let me know.

I'll investigate why CI is failing...

Samasaur1 committed:
Switch from a Swift version check to a module import check

The check for the Swift 6.0 compiler was trying to do the same thing
(ensure that the 15.0 SDK or newer was in use before compiling code that
would error on earlier SDKs), but there were niche situations where it
would not work.

Thanks to <https://marcoeidinger.github.io/appleframeworks/>, I have
found a framework (CoreHID) that only exists in the 15.0 SDK or newer,
so I have switched to checking if that module can be imported.

Samasaur1 committed:
Use Swift's ability to create a universal binary

Samasaur1 commented:
I believe I've addressed both your concerns (and I also rebased)

mostpinkest commented on Makefile:
```diff
@@ -2,25 +2,13 @@ VERSION = $(shell cat VERSION)
-	swiftc watchid-pam-extension.swift $(DEFINES) -o $(LIBRARY_PREFIX)_x86_64.so -target x86_64-$(TARGET) -emit-library
-	swiftc watchid-pam-extension.swift $(DEFINES) -o $(LIBRARY_PREFIX)_arm64.so -target arm64-$(TARGET) -emit-library
-	lipo -create $(LIBRARY_PREFIX)_arm64.so $(LIBRARY_PREFIX)_x86_64.so -output $(LIBRARY_NAME)
+	swift build -c release --arch x86_64 --arch arm64
+	mv .build/apple/Products/Release/libpam-watchid.dylib $(LIBRARY_NAME)
```
Found a couple of issues when building with the CLT.

Support for building swift packages with the CLT (especially in older versions of macOS) seems to be quite poor. In Sonoma and later, builds are significantly longer than they previously were (about double in CI and 10x on my machine). In Ventura and earlier, they outright don't seem to work.

Likewise,

> is there any reason why we're ... not going straight to something like:
> ```
> swift build -c release --arch x86_64 --arch arm64
> ```

Turns out there is a reason not to do it like that. For whatever reason, multiarch builds don't seem to work with the CLT, instead failing with:

```
error: xcbuild executable at '/Library/Developer/SharedFrameworks/XCBuild.framework/Versions/A/Support/xcbuild' does not exist or is not executable
```

Unless I've completely overlooked something to get CLT builds working, I think it would probably be best if we continue to use the old build process when the CLT is in use, and leave the swift package for full Xcode installations and builds for Nixpkgs.

```suggestion
ifeq ($(shell [[ '$(shell xcode-select -p)' == '/Library/Developer/CommandLineTools' ]] && echo true),true)
# Legacy build
# For CLT due to poor support for building swift packages.
# Swift packages do work in macOS Sonoma and later with the CLT, but are an order of magnitude slower than Xcode. 
	swiftc Sources/pam-watchid/pam_watchid.swift -o $(LIBRARY_PREFIX)_x86_64.so -target x86_64-$(TARGET) -emit-library
	swiftc Sources/pam-watchid/pam_watchid.swift -o $(LIBRARY_PREFIX)_arm64.so -target arm64-$(TARGET) -emit-library
	lipo -create $(LIBRARY_PREFIX)_arm64.so $(LIBRARY_PREFIX)_x86_64.so -output $(LIBRARY_NAME)
else
# Swift Package Manager build
	swift build -c release --arch x86_64 --arch arm64
	mv .build/apple/Products/Release/libpam-watchid.dylib $(LIBRARY_NAME)
endif
```

`TARGET` should also be reinstated to how you had it in your previous commit.

Samasaur1 replied:
The `Makefile` actually doesn't matter at all for the Nix build, because it's going to build directly with SPM . As such, I'm happy to have the `Makefile` unconditionally use `swiftc` directly if you'd like. Let me know whether you'd prefer that and I'll make the change

mostpinkest replied:
I considered that but I’d rather still be testing the Swift Package Manager build somewhere. It’s also a much cleaner way of building in places where it’s supported.

Samasaur1 committed:
Restore legacy build in Makefile

Samasaur1 commented:
Done. I restored `TARGET` as equal to `apple-macosx10.15`, which is still slightly different than the previous value of `apple-darwin$(shell uname -r)`, but both values worked with `swiftc` in my testing (building on aarch64 for both arm64 and x86_64)

mostpinkest committed:
fix makefile

Logicer16 commented:
Perfect! Thanks again for the pr!
